Polychromatic PSF

floglass opened the pull request:
Polychromatic PSFs is an option now; both filter selection and spectral type of target are now options.
The polychromatic PSF is made with 10 wavelengths, chosen in the filter pass band and multiplied by the SED of the target.
Everything looks OK, but the execution is a bit slow (rescaling each of the 10 PSFs instead of the sum of them). I'll work on smoother execution now.
Header is created with wavelength = polychromatic, spectral type, band used, and the 10 wavelengths used in microns.

floglass committed:
header creation fixed

floglass committed:
Merge branch 'master' of /home/khabal/yapsfm with conflicts.

floglass committed:
Revert: Merge branch 'master' of /home/khabal/yapsfm with conflicts. (6b1ec7260639949839e943f74207ba9e5390fda9)

floglass committed:
reverted bad changes

floglass committed:
Add command line arguments handling

arguments are -so far- wavelength and final PSF scale.
handling done using argparse module.

floglass committed:
Polychromatic PSFs first commit

Skeleton built: new class PolyPSF and sed handling, as well as filter band pass.

floglass committed:
Polychromatic PSFs first commit

Skeleton built: new class PolyPSF and sed handling, as well as filter band pass.

floglass committed:
Polychromatic PSF handled properly

execution still slow, but code is working

floglass committed:
Header showing all wavelengths used for polychromatic PSF

ijiraq commented:
The options -w and -b should not both be allowed.  You can use the mutal-exclusion option set.

https://docs.python.org/2.7/library/argparse.html#mutual-exclusion

ijiraq commented:
Looks like the 'SED' directory will need to be added to the project:

KeyboardInterrupt
pheonix:yapsfm jjk$ python psf_model.py -s 0.1 -b Z -t m -p 1
Traceback (most recent call last):
  File "psf_model.py", line 49, in <module>
    main()
  File "psf_model.py", line 43, in main
    poly.get_sed()
  File "/Users/kavelaarsj/git/yapsfm/modules.py", line 332, in get_sed
    han = open('SED/%s' % self.sed_file)
IOError: [Errno 2] No such file or directory: 'SED/sed_m0_fe0.txt'
pheonix:yapsfm jjk$

floglass committed:
SED folder added and issues addressed

removed -p, made -w and -b mutually exclusive
and required.

ijiraq commented:
The option '--type' is required if --band is given.  Currently that is not checked for, so errors are raised when user doesn't provide. 

I'm not sure how to make --type only required if --band is given...

ijiraq commented:
Likely the --type values should all be Upper case.  And perhaps you should add the '0' to the type so that, in future we could provide more specific types?  (A0, A4, A7 etc.)? Not 100% sure about that.

ijiraq commented:
If there was a 'was' (CD Matrix values?) written to the header then the scale of the image would be available to ds9. The values of CD11/CD22 would each be the pixscale, CRPIX1/2 = 1 CRVAL1/2 = 0.0 CRTYPE1 = RA__TAN  CRTYPE2 = DEC_TAN might work.

ijiraq commented:
The count levels at the edge of the polychromatic PSF in 'Z' are significantly non-zero.  I wonder if the PSF is larger enough?

ijiraq commented on modules.py:
```diff
@@ -288,4 +304,80 @@ def rebin(coords, wavelength, size, detector_scale):
+            psf = PSF(pupil, self.scale)
+            psf.resize_psf(wavelength=wavel, size=np.shape(psf.a)[0], scale=self.scale)  # scale is supposed to be 0.01
+            # psf.save('polyPSF_%s' % wavel)  # consistency test: will save all the 10 PSFs.
+            psf._a = psf.a * self._wavelength_contributions[1][i]
```
Why is this a direct manipulation of psf._a instead of just acting on psf.a ? And why not use:

psf.a *= self.wavelength_contributions[1][i] 

I'm actually a bit confused by the way you are doing loops here.

ijiraq commented on modules.py:
```diff
@@ -288,4 +304,80 @@ def rebin(coords, wavelength, size, detector_scale):
+    def wavelength_contributions(self):
+        """
+        Computes the contributions of all 10 evenly-spaced wavelength in the filter band.
+        :return: wavelength_contributions. tuple of lists, containing the 10 wavelength and their relative contribution
```
You say you are making a 'return' but you don't actually return anything?  This appears to be an initialization of the wavelength_contributions array?

ijiraq commented on modules.py:
```diff
@@ -288,4 +304,80 @@ def rebin(coords, wavelength, size, detector_scale):
+        """
+        if self._x is not None and self._flux is not None:
+            return self._x, self._flux
+        han = open('SED/%s' % self.sed_file)
```
have a look at the 'with' process for doing this.

  with open('SED/{}'.format(self.sed_file) as han:
      data = han.readlines()

no need to close as that is handled by the with statement.

ijiraq commented on modules.py:
```diff
@@ -288,4 +304,80 @@ def rebin(coords, wavelength, size, detector_scale):
+            xy = data[i].split()
+            if xy:
+                x.append(float(xy[0]))
+                flux.append(float(xy[1]))
```
have a look at 'numpy.genfromtxt' instead of open and using your own looper.

ijiraq commented on modules.py:
```diff
@@ -288,4 +304,80 @@ def rebin(coords, wavelength, size, detector_scale):
+        if self._wavelength_contributions is not None:
+            return self._wavelength_contributions
+
+        bands = dict(Z=(0.76, 0.977), Y=(0.927, 1.192), J=(1.131, 1.454), H=(1.380, 1.774), F=(1.683, 2.000),
```
This is good, I like the use of the interp function below.  You might consider having 'bands' be a class that has attributes of 'name', 'upper' and 'lower' and a method called 'samples' or some such..

ijiraq commented on modules.py:
```diff
@@ -268,7 +285,6 @@ def resize_psf(self, wavelength=.76, size=505, scale=0.110):
         print "resizing PSF to match pixel resolution of %s''/px..." % scale
```
You are using the 'print' function to report logging messages.  Have a look at the 'logging' module and try using that instead.  Then you can have verbosity levels.

ijiraq commented on SED/sed_m5_fe4.txt:
```diff
@@ -0,0 +1,100 @@
+   0.10965     2.01e-26
```
We might modify these files to give them a simple header like ' # Wavelength Intensity'

floglass committed:
Some comments addressed.

if --band is specified but not --type, raise an error.
the code now *has* to be run with command line options, at least -w or -b to specify mono- or polychromatic.
psf._a changed to psf.a with @a.setter added.
get_sed() uses numpy.genfromtxt() instead of open()
